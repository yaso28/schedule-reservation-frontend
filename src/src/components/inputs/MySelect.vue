<template>
  <CSelect
    :id="id"
    horizontal
    :label="label"
    :is-valid="validate"
    :invalid-feedback="invalidMessage"
    :value="value"
    :options="mappedOptions"
    @update:value="setValue"
  />
</template>

<script>
export default {
  name: 'MySelect',
  mixins: [require('./input-mixin').default],
  props: {
    options: {
      type: Array,
      default: () => []
    },
    valueMapper: {
      type: Function,
      default: o => o.id
    },
    labelMapper: {
      type: Function,
      default: o => o.name
    }
  },
  computed: {
    mappedOptions () {
      const me = this
      const mapped = me.options.map(option => ({
        value: me.valueMapper(option),
        label: me.labelMapper(option)
      }))
      mapped.unshift({ value: null, label: '' })
      return mapped
    }
  }
}
</script>
