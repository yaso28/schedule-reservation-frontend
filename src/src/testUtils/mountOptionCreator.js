import { createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import routes from '@/router/routes'
import Vuex from 'vuex'
import VueI18n from 'vue-i18n'
import i18nOptions from '@/i18n/options'
import VueMeta from 'vue-meta'
import CoreuiVue from '@coreui/vue'
import icons from '@/icons'
import services from '@/services'

class MountOptionCreator {
  create (storeModules = null) {
    const localVue = createLocalVue()

    // router
    localVue.use(VueRouter)
    const router = new VueRouter({
      mode: 'history',
      base: process.env.BASE_URL,
      routes
    })

    // i18n
    localVue.use(VueI18n)
    const i18n = new VueI18n(i18nOptions)

    // vue-meta
    localVue.use(VueMeta)

    // Core UI
    localVue.use(CoreuiVue)

    const option = {
      localVue,
      router,
      i18n,
      icons,
      provide: services
    }

    // store
    if (storeModules) {
      localVue.use(Vuex)
      option.store = new Vuex.Store({
        modules: storeModules
      })
    }

    return option
  }
}

export default new MountOptionCreator()
