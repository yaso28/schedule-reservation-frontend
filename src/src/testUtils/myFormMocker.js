import MyForm from '@/models/MyForm'
jest.mock('@/models/MyForm')

class MyFormMocker {
  createMockSingleField (field, initValue) {
    const input = {}
    input[field] = initValue
    return this.createMock(input)
  }

  createMock (input) {
    const mockForm = new MyForm(input)
    mockForm.values = input
    mockForm.errors = {}
    return mockForm
  }

  async setValue (wrapper, form, field, value) {
    await wrapper.setProps({
      form: Object.assign({}, form, {
        values: Object.assign({}, form.values, { [field]: value })
      })
    })
  }

  setValueNonProps (form, field, value) {
    form.values[field] = value
  }

  async setError (wrapper, form, field, errorMessage) {
    await wrapper.setProps({
      form: Object.assign({}, form, {
        errors: Object.assign({}, form.errors, { [field]: errorMessage })
      })
    })
  }
}

export default new MyFormMocker()
