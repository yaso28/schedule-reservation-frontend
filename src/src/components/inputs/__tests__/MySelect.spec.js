import { mount } from '@vue/test-utils'
import mountOptionCreator from '@/testUtils/mountOptionCreator'
import MySelect from '../MySelect'
import myFormMocker from '@/testUtils/myFormMocker'

jest.mock('@/models/MyForm')

function getOptionWrapperArray (wrapper) {
  return wrapper.find('select').findAll('option')
}

const options = [
  { id: 1, name: 'オプション1' },
  { id: 2, name: 'オプション2' },
  { id: 3, name: 'オプション3' }
]
const selectIndex = 1
const valueAfterSelect = 1

describe('MySelect', () => {
  let form
  let field
  let wrapper
  beforeEach(() => {
    const label = 'サンプル'
    field = 'sample_field'
    form = myFormMocker.createMockSingleField(field, null)

    const mountOption = mountOptionCreator.create()
    mountOption.propsData = { label, form, field, options }
    wrapper = mount(MySelect, mountOption)
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  describe('通常', () => {
    test('表示', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('バリデーションエラー', () => {
    beforeEach(async () => {
      await myFormMocker.setError(wrapper, form, field, 'エラーです')
    })

    test('表示', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('入力', () => {
    let optionWrapperArray
    beforeEach(() => {
      optionWrapperArray = getOptionWrapperArray(wrapper)
    })

    describe('有効な値を選択', () => {
      beforeEach(async () => {
        await optionWrapperArray.at(selectIndex).setSelected()
      })

      test('MyForm.setValue()を呼出', () => {
        expect(form.setValue).toHaveBeenCalledWith(field, valueAfterSelect)
      })
    })

    describe('空を選択', () => {
      beforeEach(async () => {
        await myFormMocker.setValue(wrapper, form, field, 2)
        await optionWrapperArray.at(0).setSelected()
      })

      test('MyForm.setValue()を呼出', () => {
        expect(form.setValue).toHaveBeenCalledWith(field, null)
      })
    })
  })

  describe('オプション', () => {
    describe('custom mappers', () => {
      beforeEach(() => {
        const customOptions = [
          { key: 2, value: 'カスタム２' },
          { key: 4, value: 'カスタム４' }
        ]
        wrapper.setProps({
          options: customOptions,
          valueMapper: o => o.key,
          labelMapper: o => o.value
        })
      })

      test('表示', () => {
        expect(wrapper.element).toMatchSnapshot()
      })
    })
  })
})
