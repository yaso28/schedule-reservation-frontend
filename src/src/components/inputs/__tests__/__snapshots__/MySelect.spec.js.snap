// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MySelect オプション custom mappers 表示 1`] = `
<div
  class="form-group form-row"
  role="group"
>
  <label
    class="col-form-label col-sm-3"
    for="sample_field"
  >
     サンプル 
  </label>
  <div
    class="col-sm-9"
  >
    <!---->
    <select
      class="form-control"
      id="sample_field"
    >
      <!---->
      <option
        data-key="0"
        value=""
      >
          
      </option>
      <option
        data-key="1"
        value="2"
      >
         カスタム２ 
      </option>
      <option
        data-key="2"
        value="4"
      >
         カスタム４ 
      </option>
    </select>
    <!---->
    <!---->
    <!---->
    <!---->
  </div>
</div>
`;

exports[`MySelect バリデーションエラー 表示 1`] = `
<div
  class="form-group form-row"
  role="group"
>
  <label
    class="col-form-label col-sm-3"
    for="sample_field"
  >
     サンプル 
  </label>
  <div
    class="col-sm-9"
  >
    <!---->
    <select
      class="form-control is-invalid"
      id="sample_field"
    >
      <!---->
      <option
        data-key="0"
        value=""
      >
          
      </option>
      <option
        data-key="1"
        value="1"
      >
         オプション1 
      </option>
      <option
        data-key="2"
        value="2"
      >
         オプション2 
      </option>
      <option
        data-key="3"
        value="3"
      >
         オプション3 
      </option>
    </select>
    <!---->
    <!---->
    <div
      class="invalid-feedback"
    >
      エラーです
    </div>
    <!---->
  </div>
</div>
`;

exports[`MySelect 通常 表示 1`] = `
<div
  class="form-group form-row"
  role="group"
>
  <label
    class="col-form-label col-sm-3"
    for="sample_field"
  >
     サンプル 
  </label>
  <div
    class="col-sm-9"
  >
    <!---->
    <select
      class="form-control"
      id="sample_field"
    >
      <!---->
      <option
        data-key="0"
        value=""
      >
          
      </option>
      <option
        data-key="1"
        value="1"
      >
         オプション1 
      </option>
      <option
        data-key="2"
        value="2"
      >
         オプション2 
      </option>
      <option
        data-key="3"
        value="3"
      >
         オプション3 
      </option>
    </select>
    <!---->
    <!---->
    <!---->
    <!---->
  </div>
</div>
`;
